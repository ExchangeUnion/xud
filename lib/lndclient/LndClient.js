/* eslint-disable camelcase */

const grpc = require('grpc');
const fs = require('fs');
const assert = require('assert');

/** A class representing a client to interact with a running lnd instance. */
class LndClient {
  /**
   * Create an lnd client.
   * @param {Object} options - The lnd configuration
   */
  constructor(options) {
    const { datadir, rpcProtoPath } = options;
    assert(typeof options.datadir === 'string', 'datadir must be a string');
    assert(typeof options.rpcProtoPath === 'string', 'rpcProtoPath must be a string');

    const lndCert = fs.readFileSync(`${datadir}tls.cert`);
    const credentials = grpc.credentials.createSsl(lndCert);
    const lnrpcDescriptor = grpc.load(rpcProtoPath);
    this.lightning = new lnrpcDescriptor.lnrpc.Lightning('127.0.0.1:10009', credentials);

    const adminMacaroon = fs.readFileSync(`${datadir}admin.macaroon`);
    this.meta = new grpc.Metadata();
    this.meta.add('macaroon', adminMacaroon.toString('hex'));
  }

  /**
   * Return general information concerning the lightning node including itâ€™s identity pubkey,
   * alias, the chains it is connected to, and information concerning the number of
   * open+pending channels.
   */
  getInfo() {
    return new Promise((resolve, reject) => {
      this.lightning.getInfo({}, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Attempt to add a new invoice to the lnd invoice database.
   * @param {number} value - The value of this invoice in satoshis
   */
  addInvoice(value) {
    return new Promise((resolve, reject) => {
      this.lightning.addInvoice({
        value,
        memo: 'XU',
      }, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Pay an invoice through the Lightning Network.
   * @param {string} payment_request - An invoice for a payment within the Lightning Network.
   */
  payInvoice(payment_request) {
    return new Promise((resolve, reject) => {
      this.lightning.sendPaymentSync({ payment_request }, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }
}

module.exports = LndClient;
