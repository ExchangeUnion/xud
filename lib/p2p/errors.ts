import errorCodesPrefix from '../constants/errorCodesPrefix';

const codesPrefix = errorCodesPrefix.P2P;
const errorCodes = {
  ADDRESS_ALREADY_CONNECTED: codesPrefix.concat('.1'),
  NOT_CONNECTED: codesPrefix.concat('.2'),
};

const errors = {
  ADDRESS_ALREADY_CONNECTED: (address: string) => ({
    message: `Address (${address}) already connected`,
    code: errorCodes.ADDRESS_ALREADY_CONNECTED,
  }),
  NOT_CONNECTED: (address: string) => ({
    message: `Address (${address}) is not connected`,
    code: errorCodes.NOT_CONNECTED,
  }),
};

export { errorCodes };
export default errors;
