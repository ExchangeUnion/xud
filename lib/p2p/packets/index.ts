export { default as PacketType } from './PacketType';
export { default as Packet, PacketDirection } from './Packet';
export * from './types';
