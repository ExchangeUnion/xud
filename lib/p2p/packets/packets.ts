import PacketType from './PacketType';
import OrderPacket from './OrderPacket';

const fromRaw = (type:string, body:string) => {
  switch (type) {
    case PacketType.ORDER:
      return OrderPacket.fromRaw(body);
    default:
      return null;
  }
};

export default {
  fromRaw,
};
