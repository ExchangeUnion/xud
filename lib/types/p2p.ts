export type Address = {
  host: string;
  port: number;
};

/** Information used for connecting to a remote node. */
export type NodeConnectionInfo = {
  nodePubKey: string;
  addresses: Address[];
};

export type HandshakeState = {
  version: string;
  nodePubKey: string;
  addresses?: Address[];
  pairs: string[];
  raidenAddress?: string;
  lndbtcPubKey?: string;
  lndltcPubKey?: string;
};

export function isHandshakeState(obj: any): obj is HandshakeState {
  return obj && typeof obj.version === 'string' && typeof obj.nodePubKey === 'string' && Array.isArray(obj.addresses)
    && Array.isArray(obj.pairs);
}
