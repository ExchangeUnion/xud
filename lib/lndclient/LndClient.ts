import grpc from 'grpc';
import fs from 'fs';

import Logger from '../Logger';
import BaseClient, { ClientStatus } from '../BaseClient';
import errors from './errors';
import { LightningClient } from '../proto/lndrpc_grpc_pb';
import * as lndrpc from '../proto/lndrpc_pb';

/**
 * The configurable options for the lnd client.
 */
type LndClientConfig = {
  disable: boolean;
  certpath: string;
  macaroonpath: string;
  host: string;
  port: number;
};

interface GrpcResponse {
  toObject: Function;
}

/** A class representing a client to interact with lnd. */
class LndClient extends BaseClient {
  private lightning!: LightningClient;
  private meta!: grpc.Metadata;

  /**
   * Create an lnd client.
   * @param config The lnd configuration
   */
  constructor(config: LndClientConfig) {
    super(Logger.lnd);
    const { disable, certpath, host, port, macaroonpath } = config;

    if (disable) {
      this.setStatus(ClientStatus.DISABLED);
    } else if (!fs.existsSync(certpath)) {
      this.logger.error('could not find lnd certificate, is lnd installed?');
      this.setStatus(ClientStatus.DISABLED);
    } else if (!fs.existsSync(macaroonpath)) {
      this.logger.error('could not find lnd macaroon, is lnd installed?');
      this.setStatus(ClientStatus.DISABLED);
    } else {
      const lndCert = fs.readFileSync(certpath);
      const credentials = grpc.credentials.createSsl(lndCert);
      this.lightning = new LightningClient(`${host}:${port}`, credentials);

      const adminMacaroon = fs.readFileSync(macaroonpath);
      this.meta = new grpc.Metadata();
      this.meta.add('macaroon', adminMacaroon.toString('hex'));

      this.setStatus(ClientStatus.CONNECTION_VERIFIED); // TODO: verify connection
    }
  }

  private unaryCall = <T, U>(methodName: string, params: T): Promise<U> => {
    return new Promise((resolve, reject) => {
      if (this.isDisabled()) {
        reject(errors.LND_IS_DISABLED);
        return;
      }
      this.lightning[methodName](params, this.meta, (err: grpc.ServiceError, response: GrpcResponse) => {
        if (err) {
          reject(err);
        } else {
          resolve(response.toObject());
        }
      });
    });
  }

  /**
   * Return general information concerning the lightning node including itâ€™s identity pubkey, alias, the chains it
   * is connected to, and information concerning the number of open+pending channels.
   */
  public getInfo = (): Promise<lndrpc.GetInfoResponse.AsObject> => {
    return this.unaryCall<lndrpc.GetInfoRequest, lndrpc.GetInfoResponse.AsObject>('getInfo', new lndrpc.GetInfoRequest());
  }

  /**
   * Attempt to add a new invoice to the lnd invoice database.
   * @param value The value of this invoice in satoshis
   */
  public addInvoice = (value: number): Promise<lndrpc.AddInvoiceResponse.AsObject> => {
    const request = new lndrpc.Invoice();
    request.setValue(value);
    return this.unaryCall<lndrpc.Invoice, lndrpc.AddInvoiceResponse.AsObject>('addInvoice', request);
  }

  /**
   * Pay an invoice through the Lightning Network.
   * @param payment_request An invoice for a payment within the Lightning Network.
   */
  public payInvoice = (paymentRequest: string): Promise<lndrpc.SendResponse.AsObject> => {
    const request = new lndrpc.SendRequest();
    request.setPaymentRequest(paymentRequest);
    return this.unaryCall<lndrpc.SendRequest, lndrpc.SendResponse.AsObject>('sendPaymentSync', request);
  }

  /**
   * Get a new address for the internal lnd wallet.
   */
  public newAddress = (addressType: lndrpc.NewAddressRequest.AddressType): Promise<lndrpc.NewAddressResponse.AsObject> => {
    const request = new lndrpc.NewAddressRequest();
    request.setType(addressType);
    return this.unaryCall<lndrpc.NewAddressRequest, lndrpc.NewAddressResponse.AsObject>('newAddress', request);
  }

  /**
   * Return the total of unspent outputs for the internal lnd wallet.
   */
  public walletBalance = (): Promise<lndrpc.WalletBalanceResponse.AsObject> => {
    return this.unaryCall<lndrpc.WalletBalanceRequest, lndrpc.WalletBalanceResponse.AsObject>('walletBalance', new lndrpc.WalletBalanceRequest());
  }

  /**
   * Return the total funds available across all channels.
   */
  public channelBalance = (): Promise<lndrpc.ChannelBalanceResponse.AsObject> => {
    return this.unaryCall<lndrpc.ChannelBalanceRequest, lndrpc.ChannelBalanceResponse.AsObject>('channelBalance', new lndrpc.ChannelBalanceRequest());
  }

  /**
   * Connect to another lnd node.
   */
  public connectPeer = (pubkey: string, host: string, port: number): Promise<lndrpc.ConnectPeerResponse.AsObject> => {
    const request = new lndrpc.ConnectPeerRequest();
    const address = new lndrpc.LightningAddress();
    address.setHost(`${host}:${port}`);
    address.setPubkey(pubkey);
    request.setAddr(address);
    return this.unaryCall<lndrpc.ConnectPeerRequest, lndrpc.ConnectPeerResponse.AsObject>('connectPeer', request);
  }

  /**
   * Open a channel with a connected lnd node.
   */
  public openChannel = (node_pubkey_string: string, local_funding_amount: number): Promise<lndrpc.ChannelPoint.AsObject> => {
    const request = new lndrpc.OpenChannelRequest;
    request.setNodePubkeyString(node_pubkey_string);
    request.setLocalFundingAmount(local_funding_amount);
    return this.unaryCall<lndrpc.OpenChannelRequest, lndrpc.ChannelPoint.AsObject>('openChannelSync', request);
  }

  /**
   * List all open channels for this node.
   */
  public listChannels = (): Promise<lndrpc.ListChannelsResponse.AsObject> => {
    return this.unaryCall<lndrpc.ListChannelsRequest, lndrpc.ListChannelsResponse.AsObject>('listChannels', new lndrpc.ListChannelsRequest());
  }

  /**
   * Attempt to close an open channel.
   */
  public closeChannel = (fundingTxId: string, outputIndex: number, force: boolean): void => {
    if (this.isDisabled()) {
      throw(errors.LND_IS_DISABLED);
    }
    const request = new lndrpc.CloseChannelRequest();
    const channelPoint = new lndrpc.ChannelPoint();
    channelPoint.setFundingTxidStr(fundingTxId);
    channelPoint.setOutputIndex(outputIndex);
    request.setChannelPoint(channelPoint);
    request.setForce(force);
    this.lightning.closeChannel(request, this.meta)
      // TODO: handle close channel events
      .on('data', (message: string) => {
        this.logger.info(`closeChannel update: ${message}`);
      })
      .on('end', () => {
        this.logger.info('closeChannel ended');
      })
      .on('status', (status: string) => {
        this.logger.debug(`closeChannel status: ${JSON.stringify(status)}`);
      })
      .on('error', (error: any) => {
        this.logger.error(`closeChannel error: ${error}`);
      });
  }

  /**
   * Subscribe to events for when invoices are settled.
   */
  public subscribeInvoices = (): void => {
    if (this.isDisabled()) {
      throw(errors.LND_IS_DISABLED);
    }
    this.lightning.subscribeInvoices(new lndrpc.InvoiceSubscription(), this.meta)
      // TODO: handle invoice events
      .on('data', (message: string) => {
        this.logger.info(`invoice update: ${message}`);
      })
      .on('end', () => {
        this.logger.info('invoice ended');
      })
      .on('status', (status: string) => {
        this.logger.debug(`invoice status: ${JSON.stringify(status)}`);
      })
      .on('error', (error: any) => {
        this.logger.error(`invoice error: ${error}`);
      });
  }
}

export default LndClient;
export { LndClientConfig };
