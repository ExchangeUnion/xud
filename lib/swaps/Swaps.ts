import { SwapDealRole } from '../types/enums';
import Peer from '../p2p/Peer';
import uuidv1 from 'uuid/v1';
import * as packets from '../p2p/packets/types';
import { randomBytes, createHash } from 'crypto';
import Logger from '../Logger';
import * as lndrpc from '../proto/lndrpc_pb';
import LndClient from '../lndclient/LndClient';
import Pool from '../p2p/Pool';
import { EventEmitter } from 'events';
import { StampedPeerOrder, StampedOwnOrder } from '../types/orders';

type SwapDeal = {
  /** The role of the local node in the swap. */
  myRole: SwapDealRole;
  /** The xud node pub key of the counterparty to this swap deal. */
  peerPubKey: string;
  /** Global order id in the XU network. */
  orderId: string;
  dealId: string;
  /** The quantity of the order to execute as proposed by the taker. Negative when the taker is selling. */
  proposedQuantity: number;
  /** The accepted quantity of the order to execute as accepted by the maker. Negative when the taker is selling. */
  quantity?: number;
  /** The trading pair of the order. */
  pairId: string;
  /** The number of satoshis (or equivalent) the taker is expecting to receive. */
  takerAmount?: number;
  /** The currency the taker is expecting to receive. */
  takerCurrency: string;
  /** Taker's lnd pubkey on the taker currency's network. */
  takerPubKey: string;
  /** The number of satoshis (or equivalent) the maker is expecting to receive. */
  makerAmount?: number;
  /** The currency the maker is expecting to receive. */
  makerCurrency: string;
  /** Maker's lnd pubkey on the maker currency's network. */
  makerPubKey?: string;
  /** The hash of the preimage. */
  r_hash?: string;
  r_preimage?: string;
  createTime: number;
  executeTime?: number;
  competionTime?: number
};

interface Swaps {
  on(event: 'swap.rejected', listener: (deal: SwapDeal) => void): this;
  on(event: 'swap.accepted', listener: (deal: SwapDeal, quantity: number) => void): this;
  on(event: 'swap.completed', listener: (deal: SwapDeal) => void): this;
  on(event: 'swap.failed', listener: (deal: SwapDeal) => void): this;
  emit(event: 'swap.rejected', deal: SwapDeal): boolean;
  emit(event: 'swap.accepted', deal: SwapDeal, quantity: number): boolean;
  emit(event: 'swap.completed', deal: SwapDeal): boolean;
  emit(event: 'swap.failed', deal: SwapDeal): boolean;
}

class Swaps extends EventEmitter {
  /** A mapping containing two entries for each deal, one using dealId as a key and one using r_hash as a key. */
  private deals = new Map<string, SwapDeal>();

  constructor(private logger: Logger, private pool: Pool, private lndBtcClient: LndClient, private lndLtcClient: LndClient) {
    super();

    this.bind();
  }

  /**
   * Calculates the amount of subunits/satoshis each side of a swap should receive.
   * @param quantity the quantity of the taker's order
   * @param price the price specified by the maker order being filled
   */
  private static calculateSwapAmounts = (quantity: number, price: number) => {
    let takerAmount: number;
    let makerAmount: number;
    // TODO: use configurable amount of subunits/satoshis per token for each currency
    if (quantity > 0) {
      // taker is buying the base currency
      takerAmount = Math.round(quantity * 100000000);
      makerAmount = Math.round(quantity * price * 100000000);
    } else {
      // taker is selling the base currency
      takerAmount = Math.round(quantity * price * -100000000);
      makerAmount = Math.round(quantity * -100000000);
    }

    return { takerAmount, makerAmount };
  }

  private bind() {
    this.pool.on('packet.dealRequest', this.handleDealRequest);
    this.pool.on('packet.dealResponse', this.handleDealResponse);
    this.pool.on('packet.swapRequest', this.handleSwapRequest);
    this.pool.on('packet.swapResponse', this.handleSwapResponse);
  }

  /**
   * Gets a deal by its dealId or r_hash value.
   * @param dealIdOrHash The dealId or r_hash value of the deal to get.
   * @returns A deal if one is found, otherwise undefined.
   */
  public getDeal = (dealIdOrHash: string): SwapDeal | undefined => {
    return this.deals.get(dealIdOrHash);
  }

  public addDeal = (deal: SwapDeal) => {
    this.deals.set(deal.dealId, deal);
    if (deal.r_hash) {
      this.deals.set(deal.r_hash, deal);
    }
  }

  public removeDeal = (deal: SwapDeal) => {
    this.deals.delete(deal.dealId);
    if (deal.r_hash) {
      this.deals.delete(deal.r_hash);
    }
  }

  /**
   * Begins a swap to fill an order by sending a [[DealRequestPacket]] to the maker.
   * @param maker the remote maker order we are filling
   * @param taker our local taker order
   * @returns the deal id for the swap
   */
  public beginSwap = (maker: StampedPeerOrder, taker: StampedOwnOrder) => {
    const peer = this.pool.getPeer(maker.peerPubKey);

    const [baseCurrency, quoteCurrency] = maker.pairId.split('/');
    const dealId = uuidv1();

    let takerCurrency: string;
    let makerCurrency: string;
    if (taker.quantity > 0) {
      // we are buying the base currency
      takerCurrency = baseCurrency;
      makerCurrency = quoteCurrency;
    } else {
      // we are selling the base currency
      takerCurrency = quoteCurrency;
      makerCurrency = baseCurrency;
    }
    const { takerAmount, makerAmount } = Swaps.calculateSwapAmounts(taker.quantity, maker.price);

    const dealRequestBody: packets.DealRequestPacketBody = {
      dealId,
      takerCurrency,
      makerCurrency,
      takerAmount,
      makerAmount,
      takerPubKey: takerCurrency === 'BTC' ? this.pool.handshakeData.lndbtcPubKey! : this.pool.handshakeData.lndltcPubKey!,
      orderId: maker.id,
      pairId: maker.pairId,
      proposedQuantity: taker.quantity,
    };

    const deal: SwapDeal = {
      ...dealRequestBody,
      myRole : SwapDealRole.Taker,
      peerPubKey: this.pool.handshakeData.nodePubKey,
      createTime: Date.now(),
    };

    this.addDeal(deal);
    this.logger.debug('Initiating swap deal: ' + JSON.stringify(deal));

    const packet = new packets.DealRequestPacket(dealRequestBody);

    peer.sendPacket(packet);
    return dealId;
  }

  public executeSwap = (takerCurrency: string, makerCurrency: string, takerPubKey: string,
  takerAmount: number, makerAmount: number, peer: Peer) => {
    // TODO: Remove this method after we remove Service.executeSwap()
    const dealRequestBody: packets.DealRequestPacketBody = {
      takerCurrency,
      makerCurrency,
      takerPubKey,
      takerAmount,
      makerAmount,
      dealId: uuidv1(),
      pairId: '',
      orderId: '',
      proposedQuantity: 0,
    };

    const deal: SwapDeal = {
      ...dealRequestBody,
      myRole : SwapDealRole.Taker,
      peerPubKey: peer.nodePubKey!,
      createTime: Date.now(),
    };

    const packet = new packets.DealRequestPacket(dealRequestBody);
    peer.sendPacket(packet);

    this.logger.debug('New swap deal: ' + JSON.stringify(deal));
    return deal;
  }

  /**
   * Handles a request from a peer to create a swap deal. Creates a preimage to be used for the swap, creates a deal,
   * and stores the deal in the local collection of deals. Responds to the peer with a deal response packet.
   */
  private handleDealRequest = (requestPacket: packets.DealRequestPacket, peer: Peer)  => {
    const requestBody = requestPacket.body;
    if (!requestBody) {
      return;
    }

    /** The lnd pub key to use for the local (maker) side of the deal */
    let makerPubKey: string | undefined;
    switch (requestBody.makerCurrency){
      case 'BTC':
        makerPubKey = this.pool.handshakeData.lndbtcPubKey;
        break;
      case 'LTC':
        makerPubKey = this.pool.handshakeData.lndltcPubKey;
        break;
    }

    let responsePacket: packets.DealResponsePacket;
    if (!makerPubKey) {
      // we don't support this currency, respond with an empty packet to reject the deal
      // TODO: check if we support taker currency as well
      responsePacket = new packets.DealResponsePacket({ dealId: requestBody.dealId, quantity: 0 }, requestPacket.header.id);
    } else {
      // accept the deal
      const r_preimage = randomBytes(32);
      const hash = createHash('sha256');
      const r_hash = hash.update(r_preimage).digest('hex');

      // TODO: we always accept proposed quantity currently, put in checks to make sure we have the full amount available
      const quantity = requestBody.proposedQuantity;

      const deal: SwapDeal = {
        ...requestBody,
        makerPubKey,
        quantity,
        r_hash,
        r_preimage: r_preimage.toString('hex'),
        peerPubKey: peer.nodePubKey!,
        myRole: SwapDealRole.Maker,
        createTime: Date.now(),
      };

      this.addDeal(deal);
      this.logger.debug('[SWAP] New deal: ' + JSON.stringify(deal));

      const responseBody: packets.DealResponsePacketBody = {
        makerPubKey,
        quantity,
        r_hash,
        dealId: requestBody.dealId,
      };

      responsePacket = new packets.DealResponsePacket(responseBody, requestPacket.header.id);
    }
    this.logger.debug('Sending deal response to peer: ' + JSON.stringify(responsePacket.body));
    peer.sendPacket(responsePacket);
  }

  /**
   * Handles a response from a peer to confirm a swap deal and updates the deal. If the deal is
   * accepted, sends a [[SwapRequestPacket]] to initiate the deal.
   */
  private handleDealResponse = (responsePacket: packets.DealResponsePacket, peer: Peer): void  => {
    const { dealId, quantity, r_hash, makerPubKey } = responsePacket.body!;
    const deal = this.getDeal(dealId);
    if (!deal) {
      // TODO: add error handling for an unknown deal id
      this.logger.warn(`received deal response for unrecognized deal ${dealId}`);
      return;
    }

    if (!r_hash) {
      // Our deal request doesn't have a payment hash and therefore was rejected
      this.removeDeal(deal);
      this.emit('swap.rejected', deal);
      return;
    }

    if (quantity) {
      // TODO: require a non-zero quantity value on accepted deal responses
      if (quantity > deal.proposedQuantity) {
        // TODO: this should not happen, abort deal and penalize peer
      } else if (quantity < deal.proposedQuantity) {
        // TODO: the maker accepted only part of our deal request, adjust the deal amounts with Swaps.calculateSwapAmounts
      }
      this.emit('swap.accepted', deal, quantity);
    }
    deal.makerPubKey = makerPubKey;
    deal.r_hash = r_hash;
    this.deals.set(r_hash, deal);
    this.logger.debug('[SWAP] Updated deal: ' + JSON.stringify(deal));

    const swapRequestBody: packets.SwapRequestPacketBody = { dealId };
    const swapRequestPacket = new packets.SwapRequestPacket(swapRequestBody);

    this.logger.debug('Sending swap request to peer: ' + JSON.stringify(swapRequestBody));

    peer.sendPacket(swapRequestPacket);
  }

  /**
   * Handles a request from a peer to perform a swap. Performs the actual swap via lnd and sends
   * a [[SwapResponsePacketBody]] containing the preimage after it's complete.
   */
  private handleSwapRequest = async (requestPacket: packets.SwapRequestPacket, peer: Peer) => {
    const { dealId } = requestPacket.body!;
    let responseBody: packets.SwapResponsePacketBody = { dealId };
    const deal = this.getDeal(dealId);
    if (deal && deal.r_hash) {
      // we know this deal, proceed with the swap
      let cmdLnd: LndClient;

      switch (deal.makerCurrency) {
        case 'BTC':
          cmdLnd =  this.lndLtcClient;
          break;
        case 'LTC':
          cmdLnd = this.lndBtcClient;
          break;
        default:
          // TODO: handle unsupported currency
          return;
      }

      const request = new lndrpc.SendRequest();
      request.setAmt(deal.takerAmount!);
      request.setDestString(deal.takerPubKey);
      request.setPaymentHashString(deal.r_hash);

      // TODO: use timeout on call

      try {
        const sendPaymentResponse = await cmdLnd.sendPaymentSync(request);
        if (sendPaymentResponse.getPaymentError()) {
          throw sendPaymentResponse.getPaymentError();
        }

        // swap succeeded!
        const r_preimage = Buffer.from(sendPaymentResponse.getPaymentPreimage_asB64(), 'base64').toString('hex');
        this.logger.debug(`swap succeeded! ${r_preimage}`);
        this.emit('swap.completed', deal);
        responseBody = {
          r_preimage,
          dealId,
        };
      } catch (err) {
        this.logger.error(`Got exception from sendPaymentSync ${JSON.stringify(request.toObject())}`, err);
        this.emit('swap.failed', deal);
      }
    } else {
      // this is not a valid deal
      this.logger.warn(`received swap request for unrecognized deal ${dealId}`);
      // TODO: respond with a failure message, possibly penalize peer
    }

    const responsePacket = new packets.SwapResponsePacket(responseBody, requestPacket.header.id);

    this.logger.debug('Sending swap response to peer: ' + JSON.stringify(responseBody));

    peer.sendPacket(responsePacket);
  }

  private handleSwapResponse = (response: packets.SwapResponsePacket): void  => {
    // TODO: Remove this method/packet after we detect successful swap on invoice event
    const { dealId, r_preimage } = response.body!;
    const deal = this.getDeal(dealId);
    if (!deal) {
      this.logger.warn(`received swap response for unknown deal ${dealId}`);
      return;
    }
    if (!r_preimage) {
      this.emit('swap.failed', deal);
      return;
    }

    deal.r_preimage = r_preimage;
    this.emit('swap.completed', deal);
    this.logger.debug('[SWAP] Swap completed. preimage = ' + r_preimage);
  }
}

export default Swaps;
export { SwapDeal };
