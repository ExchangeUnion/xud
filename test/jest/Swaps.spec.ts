import LndClient from '../../lib/lndclient/LndClient';
import { SwapRequestPacketBody } from '../../lib/p2p/packets';
import Swaps from '../../lib/swaps/Swaps';

describe('Swaps', () => {
  describe('calculateLockBuffer', () => {
    const ltcMinutesPerBlock = LndClient['MINUTES_PER_BLOCK_BY_CURRENCY']['LTC'];
    const btcMinutesPerBlock = LndClient['MINUTES_PER_BLOCK_BY_CURRENCY']['BTC'];
    const ethMinutesPerBlock = 0.25;

    test('it calculates a lock buffer with BTC first leg and LTC second leg', async () => {
      expect(Swaps['calculateLockBuffer'](1152, ltcMinutesPerBlock, btcMinutesPerBlock)).toMatchSnapshot();
    });

    test('it calculates a lock buffer with LTC first leg and BTC second leg', async () => {
      expect(Swaps['calculateLockBuffer'](80, btcMinutesPerBlock, ltcMinutesPerBlock)).toMatchSnapshot();
    });

    test('it calculates a lock buffer with BTC first leg and WETH second leg', async () => {
      expect(Swaps['calculateLockBuffer'](100, ethMinutesPerBlock, btcMinutesPerBlock)).toMatchSnapshot();
    });

    test('it calculates a lock buffer with WETH first leg and BTC second leg', async () => {
      expect(Swaps['calculateLockBuffer'](80, btcMinutesPerBlock, ethMinutesPerBlock)).toMatchSnapshot();
    });
  });

  describe('validateSwapRequest', () => {
    const quantity = 1000000;
    const takerCltvDelta = 144;
    const orderId = 'f8a85c66-7e73-43cd-9ac4-176ff4cc28a8';
    const rHash = '62c8bbef4587cff4286246e63044dc3e454b5693fb5ebd0171b7e58644bfafe2';

    const swapRequest: SwapRequestPacketBody = {
      takerCltvDelta,
      orderId,
      rHash,
      proposedQuantity: quantity,
      pairId: 'LTC/BTC',
    };

    it('should validate a good swap request', () => {
      expect(Swaps.validateSwapRequest(swapRequest)).toEqual(true);
    });

    it('should flag a swap request with a non-positive proposed quantity', () => {
      expect(Swaps.validateSwapRequest({ ...swapRequest, proposedQuantity: 0 })).toEqual(false);
      expect(Swaps.validateSwapRequest({ ...swapRequest, proposedQuantity: -1 })).toEqual(false);
    });

    it('should flag a swap request with an rHash that is not 64 characters', () => {
      expect(Swaps.validateSwapRequest({ ...swapRequest, rHash: 'notavalidhash' })).toEqual(false);
    });
  });
});
