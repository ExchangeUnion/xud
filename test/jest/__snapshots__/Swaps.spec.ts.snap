// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swaps calculateLockBuffer it calculates a lock buffer with BTC first leg and LTC second leg 1`] = `389`;

exports[`Swaps calculateLockBuffer it calculates a lock buffer with BTC first leg and WETH second leg 1`] = `12`;

exports[`Swaps calculateLockBuffer it calculates a lock buffer with LTC first leg and BTC second leg 1`] = `542`;

exports[`Swaps calculateLockBuffer it calculates a lock buffer with WETH first leg and BTC second leg 1`] = `4854`;
